query/define.rs

```diff
@@ -363,151 +516,442 @@ fn define_capabilities_relates(
+                    .map_err(|source| DefineError::CreateRelates { source, relates: relates.to_owned() })?;
+                (relates, true)
+            }
+            DefinitionStatus::ExistsSame(Some((existing_relates, _))) => (existing_relates, false),
+            DefinitionStatus::ExistsSame(None) => unreachable!("Existing relates concept expected"),
+            DefinitionStatus::ExistsDifferent((existing_relates, existing_ordering)) => {
+                return Err(DefineError::RelatesAlreadyDefinedButDifferent {
                     label: label.clone(),
+                    declaration: capability.to_owned(),
+                    ordering,
+                    existing_relates,
                     existing_ordering,
-                    new_ordering: ordering,
-                })?;
+                })
             }
-            existing_relates
-        } else {
-            relation_type
-                .create_relates(snapshot, type_manager, thing_manager, role_label.name.as_str(), ordering, None) // TODO: Provide card in define instead of None
-                .map_err(|source| DefineError::CreateRelates { source, relates: relates.to_owned() })?
         };
-        // Handle annotations
-        for typeql_annotation in &capability.annotations {
-            let annotation =
-                translate_annotation(typeql_annotation).map_err(|source| DefineError::LiteralParseError { source })?;
-            let relates_annotation = RelatesAnnotation::try_from(annotation.clone())
-                .map_err(|source| DefineError::IllegalAnnotation { source })?;
-            created
-                .set_annotation(snapshot, type_manager, thing_manager, relates_annotation)
-                .map_err(|source| DefineError::SetAnnotation { label: label.clone(), source, annotation })?
+
+        define_relates_annotations(
+            snapshot,
+            type_manager,
+            thing_manager,
+            &label,
+            defined.clone(),
+            &capability,
+            is_new,
+        )?;
+    }
+    Ok(())
+}
+
+fn define_relates_annotations<'a>(
+    snapshot: &mut impl WritableSnapshot,
+    type_manager: &TypeManager,
+    thing_manager: &ThingManager,
+    relation_label: &Label<'a>,
+    relates: Relates<'a>,
+    typeql_capability: &TypeQLCapability,
+    is_new: bool,
+) -> Result<(), DefineError> {
+    for typeql_annotation in &typeql_capability.annotations {
+        let annotation =
+            translate_annotation(typeql_annotation).map_err(|source| DefineError::LiteralParseError { source })?;
+        let converted_for_relates = capability_convert_and_validate_annotation_definition_need(
+            snapshot,
+            type_manager,
+            &relation_label,
+            relates.clone(),
+            annotation.clone(),
+        );
+        match converted_for_relates {
+            Ok(Some(relates_annotation)) => {
+                // New relates should set Cardinality on initialization
```

Comment:
Will be removed in the next PR when the init_cardinality is removed from the `create_relates` API!